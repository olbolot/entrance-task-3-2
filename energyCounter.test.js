const fs = require('fs');
const energyCounter = require('./energyCounter');
const ERRORS = require('./energyCounter');

let GetInitDataFromFile = function(file) {
	return JSON.parse(fs.readFileSync(file, 'utf8'));
}

test('Initial Yandex input data', () => {
	let inputData = GetInitDataFromFile("./datasets/yandex-input-data.json");
	expect(energyCounter(inputData)).toMatchSnapshot();
});

test('Device that consumes more then possible', () => {
	let inputData = GetInitDataFromFile("./datasets/impossible-power.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.MaxPowerExceededForDeviceRegEx);
});

test('Empty input data', () => {
	let inputData = GetInitDataFromFile("./datasets/empty-json.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.RatesNotDefined);
});

test('Unknown modes handling', () => {
	let inputData = GetInitDataFromFile("./datasets/unexpected-mode.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.UnexpectedModeSpecifiedForDeviceRegEx);
});

test('Rate value isNaN', () => {
	let inputData = GetInitDataFromFile("./datasets/rates-1.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.RateValueIsNaN);
});

test('Missing some hours rates', () => {
	let inputData = GetInitDataFromFile("./datasets/rates-4.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.UnspecifiedHourRate);
});

test('Rate does not have value', () => {
	let inputData = GetInitDataFromFile("./datasets/rates-2.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.MissingSomeRateProperties);
});

test('Rate to isNaN', () => {
	let inputData = GetInitDataFromFile("./datasets/rates-3.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.RateFromToIsNaN);
});

test('Rate intersection test #1', () => {
	let inputData = GetInitDataFromFile("./datasets/rates-intersect-1.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.RatesIntersects);
});

test('Rate intersection test #2', () => {
	let inputData = GetInitDataFromFile("./datasets/rates-intersect-2.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.RatesIntersects);
});

test('Rate intersection test #3', () => {
	let inputData = GetInitDataFromFile("./datasets/rates-intersect-3.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.RatesIntersects);
});

test('Device that wants to run for 26 hours', () => {
	let inputData = GetInitDataFromFile("./datasets/26-hour-running-device.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.DeviceCannotRunMoreThen24Hours);
});

test('Not enough power for hungry devices', () => {
	let inputData = GetInitDataFromFile("./datasets/hungry-devices.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.CannotFindHoursForRunningDeviceRegEx);
});

test('Not enough power for mining rigs', () => {
	let inputData = GetInitDataFromFile("./datasets/mining-rigs.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.CannotFindHoursForRunningDeviceRegEx);
});

test('Device running more then hour avaliable in selected mode', () => {
	let inputData = GetInitDataFromFile("./datasets/very-night-dish-wash.json");
	expect(() => energyCounter(inputData)).toThrow(ERRORS.RequestedTooMuchWorkingHoursForDeviceRegEx);
});