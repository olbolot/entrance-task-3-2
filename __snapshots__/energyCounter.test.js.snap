// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial Yandex input data 1`] = `
Object {
  "consumedEnergy": Object {
    "devices": Array [],
    "value": 38.939,
  },
  "schedule": Object {
    "0": Array [
      "F972B82BA56A70CC579945773B6866FB",
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
      "7D9DC84AD110500D284B33C82FE6E85E",
    ],
    "1": Array [
      "F972B82BA56A70CC579945773B6866FB",
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "10": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "11": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "12": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "13": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "14": Array [
      "C515D887EDBBE669B2FDAC62F571E9E9",
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "15": Array [
      "C515D887EDBBE669B2FDAC62F571E9E9",
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "16": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "17": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "18": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "19": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "2": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "20": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "21": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "22": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "23": Array [
      "F972B82BA56A70CC579945773B6866FB",
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "3": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "4": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "5": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "6": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "7": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "8": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
    "9": Array [
      "02DDD23A85DADDD71198305330CC386D",
      "1E6276CC231716FE8EE8BC908486D41E",
    ],
  },
}
`;
